File: customfit-flutter-client-sdk/lib/src/core/model/evaluation_context.dart
import 'context_type.dart';

/// An evaluation context that can be used for targeting
class EvaluationContext {
  /// The context type
  final ContextType type;

  /// Key identifying this context
  final String key;

  /// Name of this context (optional)
  final String? name;

  /// Properties associated with this context
  final Map<String, dynamic> properties;

  /// Private attributes that should not be sent to analytics
  final List<String> privateAttributes;

  /// Constructor
  EvaluationContext({
    required this.type,
    required this.key,
    this.name,
    this.properties = const {},
    this.privateAttributes = const [],
  });

  /// Creates an EvaluationContext from a map representation
  factory EvaluationContext.fromMap(Map<String, dynamic> map) {
    return EvaluationContext(
      type: ContextType.values.firstWhere(
        (e) => e.toString().split('.').last == map['type'],
        orElse: () => ContextType.custom,
      ),
      key: map['key'] as String,
      name: map['name'] as String?,
      properties: (map['properties'] as Map<String, dynamic>?) ?? {},
      privateAttributes: (map['private_attributes'] as List<dynamic>?)
              ?.map((e) => e as String)
              .toList() ??
          [],
    );
  }

  /// Converts the evaluation context to a map for serialization
  Map<String, dynamic> toMap() {
    return {
      'type': type.toString().split('.').last,
      'key': key,
      if (name != null) 'name': name,
      'properties': properties,
      'private_attributes': privateAttributes,
    };
  }
}

File: customfit-kotlin-client-sdk/src/main/kotlin/customfit/ai/kotlinclient/core/model/EvaluationContext.kt
package customfit.ai.kotlinclient.core.model

import customfit.ai.kotlinclient.serialization.MapSerializer
import kotlinx.serialization.Contextual
import kotlinx.serialization.Serializable

/** Represents a context type for evaluation */
enum class ContextType {
    USER,
    DEVICE,
    SESSION,
    ORGANIZATION,
    CUSTOM
}

/** An evaluation context that can be used for targeting */
@Serializable
data class EvaluationContext(
        /** The context type */
        val type: ContextType,

        /** Key identifying this context */
        val key: String,

        /** Name of this context (optional) */
        val name: String? = null,

        /** Properties associated with this context */
        @Contextual
        @Serializable(with = MapSerializer::class) 
        val properties: Map<String, @Contextual Any> = emptyMap(),

        /** Private attributes that should not be sent to analytics */
        val privateAttributes: List<String> = emptyList()
) {
    /** Convert to a map for API requests */
    fun toMap(): Map<String, Any?> =
            mapOf(
                            "type" to type.name.lowercase(),
                            "key" to key,
                            "name" to name,
                            "properties" to properties,
                            "private_attributes" to privateAttributes.takeIf { it.isNotEmpty() }
                    )
                    .filterValues { it != null }

    /** Builder for EvaluationContext */
    class Builder(private val type: ContextType, private val key: String) {
        private var name: String? = null
        private val properties = mutableMapOf<String, Any>()
        private val privateAttributes = mutableListOf<String>()

        fun withName(name: String) = apply { this.name = name }

        fun withProperties(properties: Map<String, Any>) = apply {
            this.properties.putAll(properties)
        }

        fun withProperty(key: String, value: Any) = apply { this.properties[key] = value }

        fun withPrivateAttributes(attributes: List<String>) = apply {
            this.privateAttributes.addAll(attributes)
        }

        fun addPrivateAttribute(attribute: String) = apply { this.privateAttributes.add(attribute) }

        fun build(): EvaluationContext =
                EvaluationContext(
                        type = type,
                        key = key,
                        name = name,
                        properties = properties.toMap(),
                        privateAttributes = privateAttributes.toList()
                )
    }
}
